src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs

```diff
@@ -112,5 +139,158 @@ private static bool IsFactory<TReturn>(MethodInfo? factory)
         }
+
+        // This host builder captures calls to the IHostBuilder then replays them on the application's
+        // IHostBuilder when the event fires
+        private class DeferredHostBuilder : IHostBuilder, IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object?>>
+        {
+            public IDictionary<object, object> Properties { get; } = new Dictionary<object, object>();
+
+            private readonly string[] _args;
+            private readonly Assembly _assembly;
+
+            private readonly TaskCompletionSource<IHost> _hostTcs = new();
+            private IDisposable? _disposable;
+
+            private Action<IHostBuilder> _configure = b => { };
+
+            // The amount of time we wait for the diagnostic source events to fire
+            private static readonly TimeSpan _waitTimeout = TimeSpan.FromSeconds(5);
+
+            public DeferredHostBuilder(string[] args, Assembly assembly)
+            {
+                _args = args;
+                _assembly = assembly;
+            }
+
+            public IHost Build()
+            {
+                using var subscription = DiagnosticListener.AllListeners.Subscribe(this);
+
+                // Kick off the entry point on a new thread so we don't block the current one
+                // in case we need to timeout the execution
+                var thread = new Thread(() =>
+                {
+                    try
+                    {
+                        _assembly.EntryPoint!.Invoke(null, new object[] { _args });
+
+                        // Try to set an exception if the entrypoint returns gracefully, this will force
+                        // build to throw
+                        _hostTcs.TrySetException(new InvalidOperationException("Unable to build IHost"));
+                    }
+                    catch (TargetInvocationException tie) when (tie.InnerException is StopTheHostException)
+                    {
+                        // The host was stopped by our own logic
+                    }
+                    catch (TargetInvocationException tie)
+                    {
+                        // Another exception happened, propagate that to the caller
+                        _hostTcs.TrySetException(tie.InnerException ?? tie);
+                    }
+                    catch (Exception ex)
+                    {
+                        // Another exception happened, propagate that to the caller
+                        _hostTcs.TrySetException(ex);
+                    }
+                })
+                {
+                    // Make sure this doesn't hang the process
+                    IsBackground = true
+                };
+
+                // Start the thread
+                thread.Start();
+
+                // Wait before throwing an exception
+                using var cts = new CancellationTokenSource(_waitTimeout);
+                _hostTcs.Task.Wait(cts.Token);
```

Comment:
I'm not sure if we should propagate the OperationCanceledException here or return null.